src/plugins/roam/roam.js

```diff
@@ -0,0 +1,254 @@
+                    break;
+                case "s": // back
+                    roamMove( 0, -0, 1 );
+                    break;
+                case "a": // left
+                    roamMove( -1, 0, 0 );
+                    break;
+                case "d": // right
+                    roamMove( 1, 0, 0 );
+                    break;
+                case "q": // up
+                    roamMove( 0, -1, 0 );
+                    break;
+                case "e": // down
+                    roamMove( 0, 1, 0 );
+                    break;
+
+                // Rotating part.
+                case "i": // pitch-up
+                    presentStep.dataset.rotateX = ( parseFloat( presentStep.dataset.rotateX ) || 0 ) + 30 / spf;
+                    break;
+                case "k": // pitch-down
+                    presentStep.dataset.rotateX = ( parseFloat( presentStep.dataset.rotateX ) || 0 ) - 30 / spf;
+                    break;
+                case "j": // roll-left
+                    presentStep.dataset.rotateZ = ( parseFloat( presentStep.dataset.rotateZ ) || 0 ) - 30 / spf;
+                    break;
+                case "l": // roll-right
+                    presentStep.dataset.rotateZ = ( parseFloat( presentStep.dataset.rotateZ ) || 0 ) + 30 / spf;
+                    break;
+                case "u": // yaw-left
+                    presentStep.dataset.rotateY = ( parseFloat( presentStep.dataset.rotateY ) || 0 ) + 30 / spf;
+                    break;
+                case "o": // yaw-right
+                    presentStep.dataset.rotateY = ( parseFloat( presentStep.dataset.rotateY ) || 0 ) - 30 / spf;
+                    break;
+                default:
+                    break;
+            }
+        } );
+
+        // Everything's done.
+        // Just leave the impress to update the "canvas" and we've finished the work for one frame!
+        // As there no more frame between present one and the comming one, duration should be 0.
+        api.goto( presentStep.id, 0 );
+    }
+
+    function eventHandler( event ) {
+        if ( event.type === "keydown" && roamKeys.includes( event.key ) && !roams.keys.includes( event.key ) ) {
+            roams[ event.key ].start = -1;
+            roams.keys += event.key;
+        } else if ( event.type === "keyup" && roams.keys.includes( event.key ) ) {
+            roams.keys = roams.keys.replaceAll( event.key, "" );
+        }
+    }
+
+    // Init `roams.keys` with `""`, which means no key's pressed down.
+    roams.keys = "";
+    Array.from( roamKeys ).forEach( ( key ) => {
+        roams[ key ] = {
+            start: undefined
+        };
+    } );
+
+    // Request to be called before next frame renders.
+    // And every single frame after is going to be called in the `roamAnimationFrame()` function.
+    window.requestAnimationFrame( roamAnimationFrame );
+
+    // Make the keys controlling.
+    document.addEventListener( "keydown", eventHandler );
```

Comment:
You should add event listeners via the gc helper. This will keep a reference to them and remove them when `impress().tear()` is called.

https://github.com/impress/impress.js/blob/master/src/lib/gc.js

For example:
https://github.com/impress/impress.js/blob/4c9d2e38aaf1ea038118d4ff6549609df6f1fab3/src/plugins/blackout/blackout.js#L93

You also need to do these initializations from an event listener listening to `impress:init`.

Reply:
OK. I'll fix them. Thanks for mentioning.